exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+	}
+
+	r.scheme = mgr.GetScheme()
+	return nil
+}
+
+func (r *ClusterResourceSetReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.Background()
+
+	// Fetch the ClusterResourceSet instance.
+	clusterResourceSet := &addonsv1.ClusterResourceSet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, clusterResourceSet); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(clusterResourceSet, r.Client, patch.WithStatusObservedGeneration{})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSet object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSet); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Handle normal reconciliation loop.
+	return r.reconcile(ctx, clusterResourceSet)
+}
+
+// reconcile handles cluster reconciliation.
+func (r *ClusterResourceSetReconciler) reconcile(ctx context.Context, clusterResourceSet *addonsv1.ClusterResourceSet) (ctrl.Result, error) {
+	logger := r.Log.WithValues("ClusterResourceSet", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	clusters, err := r.getClustersByClusterResourceSetSelector(ctx, clusterResourceSet)
+	if err != nil {
+		logger.Error(err, "Failed fetching clusters that matches ClusterResourceSet labels", "ClusterResourceSet", clusterResourceSet.Name)
+		return ctrl.Result{}, err
+	}
+
+	for _, cluster := range clusters {
+		if err := r.ApplyClusterResourceSet(ctx, cluster, clusterResourceSet); err != nil {
+			// NOTE: The reason of just logging the error if applying resources are failed is to avoid retries in case resources are missing.
+			logger.Error(err, "Failed applying resources to cluster", "Cluster", cluster.Name)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// getClustersByClusterResourceSetSelector fetches Clusters matched by the ClusterResourceSet's label selector that are in the same namespace as the ClusterResourceSet object.
+func (r *ClusterResourceSetReconciler) getClustersByClusterResourceSetSelector(ctx context.Context, clusterResourceSet *addonsv1.ClusterResourceSet) ([]*clusterv1.Cluster, error) {
+	clusterList := &clusterv1.ClusterList{}
+	if err := r.Client.List(ctx, clusterList, client.InNamespace(clusterResourceSet.Namespace), client.MatchingLabels(clusterResourceSet.Spec.ClusterSelector.MatchLabels)); err != nil {
+		return nil, errors.Wrap(err, "failed to list clusters")
+	}
+
+	clusters := []*clusterv1.Cluster{}
+	for i := range clusterList.Items {
+		m := &clusterList.Items[i]
```

Comment:
You don't have to change this, but when I see `m` I usually think of a `Machine`. Would `c` work here instead?